nexus/src/db/collection_attach.rs

```diff
@@ -0,0 +1,1480 @@
+    where
+        // We require this bound to ensure that "Self" is runnable as query.
+        Self: query_methods::LoadQuery<
+            'static,
+            DbConnection,
+            RawOutput<ResourceType, C>,
+        >,
+    {
+        self.get_result::<RawOutput<ResourceType, C>>(conn)
+            .map_err(AttachError::DatabaseError)
+            .and_then(Self::parse_result)
+    }
+
+    fn parse_result<E>(
+        result: RawOutput<ResourceType, C>,
+    ) -> Result<(C, ResourceType), AttachError<ResourceType, C, E>> {
+        let (
+            attached_count,
+            collection_before_update,
+            resource_before_update,
+            resource_after_update,
+        ) = result;
+
+        let collection_before_update = collection_before_update
+            .ok_or_else(|| AttachError::CollectionNotFound)?;
+
+        let resource_before_update = resource_before_update
+            .ok_or_else(|| AttachError::ResourceNotFound)?;
+
+        match resource_after_update {
+            Some(resource) => Ok((collection_before_update, resource)),
+            None => Err(AttachError::NoUpdate {
+                attached_count,
+                resource: resource_before_update,
+                collection: collection_before_update,
+            }),
+        }
+    }
+}
+
+type SelectableSqlType<Q> =
+    <<Q as diesel::Selectable<Pg>>::SelectExpression as Expression>::SqlType;
+
+impl<ResourceType, V, C> Query
+    for AttachToCollectionStatement<ResourceType, V, C>
+where
+    ResourceType: Selectable<Pg>,
+    C: DatastoreAttachTarget<ResourceType>,
+{
+    type SqlType = (
+        // The number of resources attached to the collection before update.
+        BigInt,
+        // If the collection exists, the value before update.
+        Nullable<SelectableSqlType<C>>,
+        // If the resource exists, the value before update.
+        Nullable<SelectableSqlType<ResourceType>>,
+        // If the resource was updated, the new value.
+        Nullable<SelectableSqlType<ResourceType>>,
+    );
+}
+
+impl<ResourceType, V, C> RunQueryDsl<DbConnection>
+    for AttachToCollectionStatement<ResourceType, V, C>
+where
+    ResourceType: Selectable<Pg>,
+    C: DatastoreAttachTarget<ResourceType>,
+{
+}
+
+/// This implementation uses a CTE which attempts to do the following:
+///
+/// 1. (collection_by_id, resource_by_id): Identify if the collection and
+///    resource objects exist at all.
+/// 2. (resource_count): Identify if the number of resources already attached to
+///    the collection exceeds a threshold.
+/// 3. (collection_info, resource_info): Checks for arbitrary user-provided
+///    constraints on the collection and resource objects.
+/// 4. (do_update): IFF all previous checks succeeded, make a decision to perfom
+///    an update.
+/// 5. (updated_resource): Apply user-provided updates on the resource -
+///    presumably, setting the collection ID value.
```

Comment:
This hopefully describes the meat of the CTE.

The TL;DR: We want to check a few things, *then* do an update if all those checks pass, and we want it to be atomic, even though multiple tables are involved.